File: adventjs/2025/15/solution.js
/**
 * @param {Array<Object>} data - The data to draw the table
 * @param {string} sortBy - The field to sort the table
 * @returns {string}
 */
function drawTable(data, sortBy) {
  if (!data.length) return ""

  const compare = (a, b) => {
    const va = a[sortBy],
      vb = b[sortBy]
    if (typeof va === "number" && typeof vb === "number") return va - vb
    return String(va).localeCompare(String(vb))
  }
  const sorted = [...data].sort(compare)

  const keys = Object.keys(data[0])
  const toStr = (v) => (v == null ? "" : String(v))
  const widths = keys.map((k) =>
    Math.max(1, ...sorted.map((r) => toStr(r[k]).length))
  )
  const sep = "+" + widths.map((w) => "-".repeat(w + 2)).join("+") + "+"
  const row = (vals) =>
    "| " + vals.map((v, i) => v.padEnd(widths[i])).join(" | ") + " |"

  return [
    sep,
    row(keys.map((_, i) => String.fromCharCode(65 + i))),
    sep,
    ...sorted.map((r) => row(keys.map((k) => toStr(r[k])))),
    sep,
  ].join("\n")
}

File: adventjs/2025/15/solution.py
def draw_table(data: list[dict], sortBy: str) -> str:
    if not data:
        return ''

    def fmt(v):
        if type(v) is bool:
            return str(v).lower()
        if v is None:
            return ''
        return str(v)

    def make_row(vals, ws):
        cells = map(lambda vw: vw[0].ljust(vw[1]), zip(vals, ws))
        return '| ' + ' | '.join(cells) + ' |'

    rows = sorted(data, key=lambda r: (type(r[sortBy]) is str, r[sortBy]))
    keys = list(data[0].keys())
    ws = [max(1, *map(lambda r: len(fmt(r[k])), rows)) for k in keys]

    sep = '+' + '+'.join('-' * (w + 2) for w in ws) + '+'
    head = make_row([chr(65 + i) for i in range(len(keys))], ws)
    body = [make_row([fmt(r[k]) for k in keys], ws) for r in rows]

    return '\n'.join([sep, head, sep] + body + [sep])
